[Camera] Add Matter Camera TCs (TC-AVSM-2.1, TC-AVSM-2.2, TC-AVSM-2.3, TC-AVSM-2.4, TC-AVSM-2.5, TC-AVSM-2.6, TC-AVSM-2.7, TC-AVSM-2.8, TC-AVSM-2.9, TC-AVSM-2.10, TC-AVSM-2.11)

chulspro opened the pull request:
### Change overview
This PR Adds the following test cases:
- TC-AVSM-2.1
- TC-AVSM-2.2
- TC-AVSM-2.3
- TC-AVSM-2.4
- TC-AVSM-2.5
- TC-AVSM-2.6
- TC-AVSM-2.7
- TC-AVSM-2.8
- TC-AVSM-2.9
- TC-AVSM-2.10
- TC-AVSM-2.11

### Testing
Build python wheel and activate venv:

. ./scripts/activate.sh
./scripts/build_python.sh -i out/python_env
source out/python_env/bin/activate

#### e.g., To run tests for TC-AVSM-2.1:

./scripts/tests/run_python_test.py --factory-reset --app <chip-camera-app> --app-args "--trace-to json:log" --script src/python_testing/TC_AVSM_2_1.py --script-args "--commissioning-method on-network --qr-code MT:-24J0AFN00KA0648G00"

Commit:
Add TC_AVSM_2_1 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_2 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_4 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_3 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_5 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add explicit precondition steps

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_6 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Fix TC_AVSM_2_5

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_7 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_8 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Restyle

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_9 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_10 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TC_AVSM_2_11 to python_testing

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

chulspro commented:
@marktrayer , @saurabhst , @yufengwangca : Please review this for AV stream management.

Commit:
Add CI test arguments

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

marktrayer commented on src/python_testing/TC_AVSM_2_2.py:
```diff
@@ -0,0 +1,196 @@
+                "DUT responds with a CONSTRAINT_ERROR status code.",
+            ),
+        ]
+
```
There's new decorators which enure this TC is only run when the cluster is present on the endpoint, otherwise what can happen is this is executed, and fails, on an EP where the cluster isn't actually present.
Suggest:
from chip.testing.matter_testing import has_cluster, run_if_endpoint_matches
...
@run_if_endpoint_matches(has_cluster(Clusters.CameraAvStreamManagement))

chulspro replied:
Thank you for your review. We fixed it with you comments "run_if_endpoint_matches"

marktrayer commented on src/python_testing/TC_AVSM_2_2.py:
```diff
@@ -0,0 +1,196 @@
+                quality=90,
+            )
+            await self.send_single_cmd(endpoint=endpoint, cmd=snpStreamAllocateCmd)
+            asserts.assert_true(False, "Unexpected success when expecting CONSTRAINT_ERROR")
```
Could the error message be more specific to the nature of the constraint error?  In this case "maxResolution" is < "minResolution"

chulspro replied:
Thank you for your comments. We fixed it with specific comments

marktrayer commented on src/python_testing/TC_AVSM_2_2.py:
```diff
@@ -0,0 +1,196 @@
+            await self.send_single_cmd(endpoint=endpoint, cmd=snpStreamAllocateCmd)
+            asserts.assert_true(False, "Unexpected success when expecting CONSTRAINT_ERROR")
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.ConstraintError, "Unexpected status returned when expecting CONSTRAINT_ERROR")
```
Could the error message be more specific as the the expected error reason?  In this case invalid quality setting.

chulspro replied:
Thank you for your comments. We added specific comment with your comments

marktrayer commented on src/python_testing/TC_AVSM_2_3.py:
```diff
@@ -0,0 +1,170 @@
+            asserts.assert_equal(e.status, Status.Success, "Unexpected error returned")
+            pass
+
+    @async_test_body
```
```suggestion
    @run_if_endpoint_matches(has_feature(Clusters.CameraAvStreamManagement,
                                                 Clusters.CameraAvStreamManagement.Bitmaps.Feature.kSnapshot)
```

chulspro replied:
Thank you for your comments. We fixed it with our new commit.

marktrayer commented on src/python_testing/TC_AVSM_2_3.py:
```diff
@@ -0,0 +1,170 @@
+
+        self.step(3)
+        try:
+            # TODO: SnapshotStreamStruct does not have WaterMarkEnabled and OSDEnabled fields. Update This.
```
Yeah .,.. that's a spec problem, needs fixing there first.  The struct definition is missing those fields, while the command does have them present.

chulspro replied:
Please fix it. Thank you.

marktrayer commented on src/python_testing/TC_AVSM_2_4.py:
```diff
@@ -0,0 +1,176 @@
+                "Verify the number of allocated snapshot streams in the list is 0.",
+            ),
+        ]
+
```
As this is the 2nd time this method has been declared local to the class it may be worth refactoring into a base class for AVSM test cases, and move this there.  See what DEM did and AVSUM did.

chulspro replied:
Thank you for your comments. We fixed it with new commit.

marktrayer commented on src/python_testing/TC_AVSM_2_5.py:
```diff
@@ -0,0 +1,232 @@
+            ),
+        ]
+
+    @async_test_body
```
```suggestion
    @run_if_endpoint_matches(has_feature(Clusters.CameraAvStreamManagement,                                         Clusters.CameraAvStreamManagement.Bitmaps.Feature.kAudio))
```

chulspro replied:
Thank you for your comments. We fixed it with our new commit.

marktrayer commented on src/python_testing/TC_AVSM_2_7.py:
```diff
@@ -0,0 +1,339 @@
+            videoStreamAllocateCmd = commands.VideoStreamAllocate(
+                streamUsage=notSupportedStreamUsage,
+                videoCodec=aRateDistortionTradeOffPoints[0].codec,
+                maxFrameRate=aVideoSensorParams.maxFPS,
```
minFrameRate is mandatory in the command, won't this fail in Python execution because it appears to be missing here?

chulspro replied:
Thank you for your comments. We fixed it with new comment

marktrayer commented on src/python_testing/TC_AVSM_2_7.py:
```diff
@@ -0,0 +1,339 @@
+            ),
+        ]
+
+    @async_test_body
```
```suggestion
    @run_if_endpoint_matches(has_feature(Clusters.CameraAvStreamManagement,                                         Clusters.CameraAvStreamManagement.Bitmaps.Feature.kVideo))
```

chulspro replied:
Thank you for your comments. We fixed it with our new commit.

marktrayer commented on src/python_testing/TC_AVSM_2_10.py:
```diff
@@ -0,0 +1,215 @@
+        asserts.assert_equal(len(aAllocatedSnapshotStreams), 1, "The number of allocated snapshot streams in the list is not 1.")
+        aStreamID = aAllocatedSnapshotStreams[0].snapshotStreamID
+
+        self.step(3)
```
Shouldn't we also check that captureSnapshotResponse.data is not None, and that the image codec and resolution are coherent?

chulspro replied:
Thank you for your comments. We fixed it. You are right.

Commit:
Fix: Review Comments

Review Comments: Add @run_if_endpoint_matches
- Add @run_if_endpoint_matches to all AVSM TCs

Review Comments: TC_AVSM_2_2
- Remove duplicate step 8
- Better error messages when expecting CONSTRAINT_ERROR

Review Comment: Create a Base class for AVSM TCs

Review Comments: TC_AVSM_2_7
- Better error message for commands expected to fail
- Add missing minFrameRate paramter to command

Review Comments: TC_AVSM_2_10
- Added checks for CaptureSnapshotResponse

Review Comments: Better error messages for all commands expected to fail

Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Add TCs that cannot be run to not_automated

This is required because some steps in the test plan cannot
be executed on the current version of camera app.
For example reading AllocatedSnapshotStreams always returns
an empty list, even if a snapshot stream has been allocated.
Similar issues exist for most of the attributes and commands.

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

marktrayer commented on src/python_testing/TC_AVSM_2_8.py:
```diff
@@ -0,0 +1,206 @@
+        vdoSupport = (aFeatureMap & cluster.Bitmaps.Feature.kVideo) > 0
+        wmarkSupport = (aFeatureMap & cluster.Bitmaps.Feature.kWatermark) > 0
+        osdSupport = (aFeatureMap & cluster.Bitmaps.Feature.kOnScreenDisplay) > 0
+        asserts.assert_true(
```
This shouldn't fail if neither watermark nor OSD are supported, it just shouldn't be run at all.  
Handle that as a decorator along the lines of:
@run_if_endpoint_matches(has_feature(Clusters.CameraAvStreamManagement,                                         Clusters.CameraAvStreamManagement.Bitmaps.Feature.kWatermark) or
                             has_feature(Clusters.CameraAvStreamManagement, Clusters.CameraAvStreamManagement.Bitmaps.Feature.kOnScreenDisplay))

chulspro replied:
Thank you for your comments. We fixed it with our new commit.

chulspro replied:
Okay, Thank you for your comments. so, we need to add WMARK or OSD in camera app or we need to remove it from this patch? Which is better?

marktrayer replied:
Have the testcase only run if WMARK or OSD are present; which is what the @run_if_endpoint I posted above does.

chulspro commented on src/python_testing/TC_AVSM_2_2.py:
```diff
@@ -0,0 +1,196 @@
+                "TH sends the SnapshotStreamAllocate command with values from step 3 except with Quality set to 101(outside of valid range).",
+                "DUT responds with a CONSTRAINT_ERROR status code.",
+            ),
+            TestStep(
```
Thank you for your comments. We fixed it. You are right.

chulspro commented on src/python_testing/TC_AVSM_2_7.py:
```diff
@@ -0,0 +1,339 @@
+            await self.send_single_cmd(endpoint=endpoint, cmd=videoStreamAllocateCmd)
+            asserts.assert_true(False, "Unexpected success when expecting CONSTRAIN_ERROR")
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.ConstraintError, "Unexpected error returned when expecting CONSTRAIN_ERROR")
```
Thank you for your comments. We fixed it with new submit

pidarped commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+            )
+            await self.send_single_cmd(endpoint=endpoint, cmd=videoStreamAllocateCmd)
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.Success, "Unexpected error returned")
```
Expected response is VideoStreamAllocateResponse with a StreamID.

chulspro replied:
Okay, Thank you for update. We had changed it to streamId

pidarped commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+                videoCodec=aRateDistortionTradeOffPoints[0].codec,
+                minFrameRate=aVideoSensorParams.maxFPS,
+                maxFrameRate=aVideoSensorParams.maxFPS,
+                minResolution=aRateDistortionTradeOffPoints[0].resolution,
```
The minResolution should be picked from minViewPort, presumably.

chulspro replied:
Okay, Thank you for review. We updated it.

pidarped commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+            videoStreamAllocateCmd = commands.VideoStreamAllocate(
+                streamUsage=aRankedStreamPriorities[0],
+                videoCodec=aRateDistortionTradeOffPoints[0].codec,
+                minFrameRate=aVideoSensorParams.maxFPS,
```
Instead of setting the minFrameRate to the sensor MaxFPS, I would suggest setting it to an acceptable minFrameRate, say 15, with a comment.

pidarped commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+                minFrameRate=aVideoSensorParams.maxFPS,
+                maxFrameRate=aVideoSensorParams.maxFPS,
+                minResolution=aRateDistortionTradeOffPoints[0].resolution,
+                maxResolution=aRateDistortionTradeOffPoints[0].resolution,
```
The maxResolution should be the VideoSensor dimensions in VideoSensorParams.

Commit:
Fix: Review Comments

- TC should not run if no support for (WMARK or OSD)
- Fix pics_guard in TC 2.1
- Assert StreamID for AllocateStream commands
- Fix VideoStreamAllocate command

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

Commit:
Update test_metadata

Signed-off-by: Subodh Singh <subodh.singh@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

chulspro commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+            )
+            await self.send_single_cmd(endpoint=endpoint, cmd=snpStreamAllocateCmd)
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.Success, "Unexpected error returned")
```
Okay, Thank you for update. We had changed it to streamId

chulspro commented on src/python_testing/TC_AVSMTestBase.py:
```diff
@@ -0,0 +1,173 @@
+            )
+            await self.send_single_cmd(endpoint=endpoint, cmd=adoStreamAllocateCmd)
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.Success, "Unexpected error returned")
```
Okay, Thank you for update. We had changed it to streamId

chulspro commented:
@marktrayer , @pidarped : Please review this again. Thank you for your time.

Commit:
Update not_automated list of test_metadata

Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

chulspro committed:
Merge branch 'master' into feature/contribution-testplans-avsm
